### PR TITLE
support multiple versions build

### PR DESCRIPTION
**Pull request by superyyrrzz**
#396 

@chenkennt @vwxyzh @ansyral @qinezh @hellosnow 

sample docfx.json:
```json
{
  "build": {
    "content": [
      {
        "files": "*.yml",
        "src": "v2.0",
        "dest": "v2.0/api",
        "version": "v2.0"
      },
      {
        "files": "*.yml",
        "src": "v2.1",
        "dest": "v2.1/api",
        "version": "v2.1"
      },
      {
        "files": "conceptual/*.md",
        "src": "v2.0",
        "dest": "v2.0",
        "version": "v2.0"
      },
      {
        "files": "conceptual/*.md",
        "src": "v2.1",
        "dest": "v2.1",
        "version": "v2.1"
      },
      {
        "files": "index.md"
      }
    ],
    "dest": "_site"
  }
} 
```

### CODE REVIEW

**Comment by vwxyzh** [comment]
👍

**Comment by hellosnow** [comment]
:shipit:

**Comment by qinezh** [comment]
Seems that we need to rebase this PR together as there're too many conflicts :joy:

File: Documentation/tutorial/docfx.exe_user_manual.md [hunk 1]
```diff
@@ -277,54 +277,45 @@ Please refer to [How to Create Custom Templates](howto_create_custom_template.md
...
 Property Name      | Description
 -------------------|----------------------------- 
 files              | **REQUIRED**. The file or file array, `glob` pattern is supported.
-name               | The folder name for the generated files.
+~~name~~           | **Obsoleted**, please use `dest`.
 exclude            | The files to be excluded, `glob` pattern is supported.
 ~~cwd~~            | **Obsoleted**, please use `src`.
 src                | Specifies the source directory. If omitted, the directory of the config file will be used. Use this option when you want to refer to files in relative folders while want to keep folder structure. e.g. set `src` to `..`.
+dest               | The folder name for the generated files.
+version            | Version name for the current file mapping. If not set, treat the current file-mapping item as in global version. Mappings with the same version name will be built together. Cross reference doesn't support cross different versions.
```
Comment (chenkennt) [inline Documentation/tutorial/docfx.exe_user_manual.md]: global version -> default version?

**Comment by qinezh** [comment]
:shipit:

File: Documentation/tutorial/docfx.exe_user_manual.md (hunk 1 above)
Comment (superyyrrzz) [inline Documentation/tutorial/docfx.exe_user_manual.md]: resolved

File: src/docfx/SubCommands/DocumentBuilderWrapper.cs [hunk 2]
```diff
@@ -225,7 +228,112 @@ private static DocumentBuildParameters ConfigToParameter(BuildJsonConfig config,
...
+                    }
+                    else
+                    {
+                        result[version] = new FileMappingParameters
+                        {
+                            Resource = new FileMapping(item)
+                        };
+                    }
+                }
+            }
```
Comment (chenkennt) [inline src/docfx/SubCommands/DocumentBuilderWrapper.cs]: These three code pieces are similar, is it possible to refactor?

Comment (superyyrrzz) [inline src/docfx/SubCommands/DocumentBuilderWrapper.cs]: Refactored. Change FileMappingParameters to a dictionary, so the property can be converted to a key as a parameter.

**Comment by ansyral** [comment]
>         }

useless?

---
Refers to: src/Microsoft.DocAsCode.Build.Engine/TemplateProcessor.cs:258 in d9b55d3. [](commit_id = d9b55d33ef23d40c6cc5ed413682658a88d186a7, deletion_comment = False)

File: src/docfx/SubCommands/DocumentBuilderWrapper.cs [hunk 3]
```diff
@@ -92,16 +93,24 @@ public static void BuildDocument(BuildJsonConfig config, TemplateManager templat
...
-                {
-                    Logger.LogWarning("No files found, nothing is to be generated");
-                    return;
-                }
-
                 using (new PerformanceScope("building documents", LogLevel.Info))
                 {
-                    builder.Build(parameters);
+                    foreach (var parameters in ConfigToParameter(config, templateManager, baseDirectory, outputDirectory))
+                    {
```
Comment (ansyral) [inline src/docfx/SubCommands/DocumentBuilderWrapper.cs]: question: seems build per version, so currently we still couldn't handle cross version reference case?

Comment (superyyrrzz) [inline src/docfx/SubCommands/DocumentBuilderWrapper.cs]: No, we doesn't it for now.

---
In reply to: [70196210](https://github.com/dotnet/docfx/pull/477#discussion_r70196210) [](ancestors = 70196210)

**Comment by superyyrrzz** [comment]
>         }

yeah.. should remove it.

---
In reply to: [231630240](https://github.com/dotnet/docfx/pull/477#issuecomment-231630240) [](ancestors = 231630240)

---
Refers to: src/Microsoft.DocAsCode.Build.Engine/TemplateProcessor.cs:258 in d9b55d3. [](commit_id = d9b55d33ef23d40c6cc5ed413682658a88d186a7, deletion_comment = False)

**Comment by ansyral** [comment]
:+1:

**Comment by chenkennt** [comment]
:shipit:

**Comment by sammychinedu2ky** [comment]
Please is there a way to build just a particular version to prevent other API versions from rebuilding? @superyyrrzz

**Pull request merged**